Include timeouts as failing status in Tester

milljm opened the pull request:
Include all the failures the StatusSystem supports, in the Tester class.

Make better use of separate OK/Fail files.

Use the json formatted output feature of qstat!

Prevent needless copies of large amounts of output data from separate output files entering into the .previous_results json file. Python needs to load this file in order to perform basic routines _not_ needing the testers output.

Closes #22710

Commit:
Include timeouts as failing status in Tester

Include all the failures the StatusSystem supports, in the Tester class.

Make better use of separate OK/Fail files.

Use the json formatted output feature of qstat!

Prevent needless copies of large amounts of output data from separate
output files entering into the .previous_results json file. Python
needs to load this file in order to perform basic routines _not_ needing
the testers output.

Closes #22710

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/1238405/) on c69bf27 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/22735/site)

This comment will be updated on new commits.

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/1238418/) on c69bf27 wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>134b8e</th>
      <th colspan=4>#22735 c69bf2</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>84.39%</td>
      <td>84.39%</td>
      <td>+0.00%</td>
      <td>-</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>80647</td>
      <td>80648</td>
      <td>+1</td>
      <td>0</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>14918</td>
      <td>14917</td>
      <td>-1</td>
      <td>0</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/22735/coverage_diff/framework)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/22735/coverage/framework)


## Modules coverage

Coverage did not change

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/22735/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/22735/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/22735/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/22735/coverage/contact)
- [`electromagnetics`](https://mooseframework.inl.gov/docs/PRs/22735/coverage/electromagnetics)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/22735/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/22735/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/22735/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/22735/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/22735/coverage/geochemistry)
- [`heat_conduction`](https://mooseframework.inl.gov/docs/PRs/22735/coverage/heat_conduction)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/22735/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/22735/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/22735/coverage/navier_stokes)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/22735/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/22735/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/22735/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/22735/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/22735/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/22735/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/22735/coverage/richards)
- [`scalar_transport`](https://mooseframework.inl.gov/docs/PRs/22735/coverage/scalar_transport)
- [`solid_properties`](https://mooseframework.inl.gov/docs/PRs/22735/coverage/solid_properties)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/22735/coverage/stochastic_tools)
- [`tensor_mechanics`](https://mooseframework.inl.gov/docs/PRs/22735/coverage/tensor_mechanics)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/22735/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/22735/coverage/xfem)

</details>



This comment will be updated on new commits.

milljm commented:
I want this to work with `--show-last-run`.

Also this reveals we need another unittest for `--show-last-run`, when the previous run was used with `--sep-files || --sep-files-ok || --sep-files-fail` arguments

GiudGiud commented:
@socratesgorilla want to review?

@loganharbour I assigned you for CCB review if you have time otherwise please re-assign

socratesgorilla commented on python/TestHarness/TestHarness.py:
```diff
@@ -794,6 +794,14 @@ def writeResults(self):
+
+                # If output has been stored in separate files, don't make additional copies by
+                # storing that data in this json results file (--pbs || --sep-files, etc options).
+                if job.getOutputFile() is not None:
+                    output = ''
+                else:
+                    output = job.getOutput()
```
Non-actionable, but just for future reference -- anything that evaluates to None in a python conditional is considered "False-y", and anything that returns a value will evaluate as True. So the pythonic way to do this is 

```suggestion
                output = '' if job.getOutputFile() else job.getOutput()
```

The exception to this is if like, getOutputFile is specifically supposed to return True or False, and you are checking if it is None, but if it isn't, the way i described is the most pythonic

You do this in other parts of the PR so you probably already know this, and it doesn't really matter anyway.

milljm replied:
I like it

socratesgorilla commented on python/TestHarness/schedulers/QueueManager.py:
```diff
@@ -41,6 +41,7 @@ def __init__(self, harness, params):
         self.options = self.harness.getOptions()
         self.__job_storage_file = self.harness.original_storage
         self.__clean_args = None
+        self.__status_system = StatusSystem()
```
Just to make sure -- this line (I believe) creates a *new* status system. Is this intentional, or is it supposed to be referencing an existing status system?

loganharbour replied:
It's a new one. But if you look at what a `StatusSystem` is - it's really just a data struct that helps you understand statuses

loganharbour commented on python/TestHarness/TestHarness.py:
```diff
@@ -794,6 +794,14 @@ def writeResults(self):
                 # Create empty key based on TestDir, or re-inialize with existing data so we can append to it
                 self.options.results_storage[job.getTestDir()] = self.options.results_storage.get(job.getTestDir(), {})
+
+                # If output has been stored in separate files, don't make additional copies by
```
I assume this is one of the cases of duplication? 

This makes me realize how badly we need more TestHarness testing and documentation :(

milljm replied:
indeed.

```python
    def writeResults(self)
```
is under much development in my PR. We do have unittesting for the sep-file(s) arguments (I think we do anyway), just not for PBS usage of sep-files. I'll have to come up with some mocking to make this work.

loganharbour commented on python/TestHarness/schedulers/QueueManager.py:
```diff
@@ -195,8 +196,11 @@ def getRunTestsCommand(self, job):
+                     os.path.join(job.getTestDir(),
+                     self.options.input_file_name),
+                     '-o', job.getTestDir(),
+                     '--sep-files'])
```
Update the comment above

loganharbour commented on python/TestHarness/schedulers/RunPBS.py:
```diff
@@ -47,7 +47,10 @@ def hasTimedOutOrFailed(self, job_data):
         if launch_id:
-            qstat_command_result = util.runCommand('qstat -xf %s' % (launch_id))
+            qstat_command_result = util.runCommand(f'qstat -xf -F json {launch_id}')
+            json_out = json.loads(qstat_command_result)
```
Beautiful json is beautiful ❤️

loganharbour commented on python/TestHarness/schedulers/RunPBS.py:
```diff
@@ -57,19 +60,22 @@ def hasTimedOutOrFailed(self, job_data):
                 return True
 
-            qstat_job_result = re.findall(r'Exit_status = (.*)', qstat_command_result)
+            job_result = job_meta.get('Exit_status', False)
+            job_output = job_meta.get('Output_Path', False)
```
Shouldn't we die more horribly if we can't find those keys? That's indicative of PBS changing

milljm replied:
I need to spend time catching up with this PR... long break, forget where I left off, that sort of thing.

We don't want to die on 'Exit_status' not being there, because this key will not exist until the job begins (if its Queued, this key does not exist). I don't know about the Output_Path. I think that _always_ exists, so... yes to dying horribly on that one.

milljm replied:
Output_Path is always present, set by the qsub script launching the job. So I removed the .get method.

Commit:
inching closer to a solution

WIP: following changes for optimizing results file.
TODO: dirty files list generation (for cleanup routines)

Commit:
WIP: current changes

Working, but still needs work...

Commit:
Allow --dry-run options

Commit:
pass -j cpus for run_tests

TODO: when all is said and done, the TestHarness sees pending tests :(

Commit:
fix conflicts

Commit:
do not read output files to screen

When using PBS, do not print failed output to screen. While we probably
should print output to the screen for failed tests, the most used cases
are those from folks running on our HPC systems (a slow filesystem). By
printing the output filename/path instead of the output, vast amounts
of time is saved by not reading sometimes-hundreds-of-megabytes-of-output
from output files. This feature is negated when using --re, --verbose,
--failed-tests options.

Remove the deprecated queue options.

Commit:
Allow verbose features, except when --quiet

Requested by Daniel Vanwasshenova

vanwdani commented:
Awesome stuff! Can't wait to use it (and maybe abuse it?)

milljm commented on python/TestHarness/schedulers/RunPBS.py:
```diff
@@ -33,21 +33,30 @@ def getBadKeyArgs(self):
+        PBS_STATUSES = { 'R' : 'RUNNING',
+                         'F' : 'FINISHED',
+                         'Q' : 'QUEUED',
+                         'E' : 'EXITING'}
```
A beautiful corner case I saw during my testing. E... PBS was finishing up (exiting). I assume this means it was busy writing output.

Rather than happenstance, I should see what other possible status PBS can have.
